SingleFile: Enable hammer servicing

swaroop-sridhar opened the pull request:
This change enables hammer servicing to work on single-file apps.

The runtime resolves assemblies in the bundle even before attempting to resolve using the TPA.
This masks any entries in the TPA that correspond to hammer-serviced DLLs.
This change fixes this problem: If a file is resolved to the servicing directory, it is disabled in the bundle directory.

Testing: Added automated tests on Unix systems, manually tested on Windows.

Fixes #36031

swaroop-sridhar committed:
SingleFile: Enable hammer servicing

This change enables hammer servicing to work on single-file apps.

The runtime resolves assemblies in the bundle even before attempting to resolve using the TPA.
This masks any entries in the TPA that correspond to hammer-serviced DLLs.
This change fixes this problem: If a file is resolved to the servicing directory, it is disabled in the bundle directory.

Testing: Added automated tests on Unix systems, manually tested on Windows.

Fixes #36031

Wraith2 commented:
What is hammer servicing?

vitek-karas commented:
@Wraith2 "hammer servicing" is a mechanism in the host which allows overriding specific assembly versions with new binaries for servicing purposes. The idea is that if there's a serious security issue the fix can be deployed by putting the fixed version of the assembly into a global location on the machine, and any application will see it there and load it in favor of the possibly vulnerable version it has in it (or in the shared framework).

This is implemented by looking for the assembly in the servicing location first, before looking in the app or in the frameworks. See the list of probing paths (order matters): https://github.com/dotnet/runtime/blob/master/docs/design/features/host-probing.md#probing-paths. The servicing paths are looked at first, and only after that the application itself is considered.

vitek-karas commented on src/installer/corehost/cli/bundle/runner.cpp:
```diff
@@ -51,7 +51,7 @@ const file_entry_t*  runner_t::probe(const pal::string_t &relative_path) const
     for (const file_entry_t& entry : m_manifest.files)
     {
-        if (pal::pathcmp(entry.relative_path(), relative_path) == 0)
+        if (entry.matches(relative_path))
```
Can you please add at least asserts into the other 'probe' and 'locate' methods that when they return an entry the entry is never disabled? Just to double check.

vitek-karas commented on src/installer/corehost/cli/deps_entry.cpp:
```diff
@@ -59,18 +59,19 @@ bool deps_entry_t::to_path(const pal::string_t& base, const pal::string_t& ietf_
     append_path(&sub_path, file_path.c_str());
 
-    if (look_in_bundle && bundle::info_t::is_single_file_bundle())
+    // Don't look_in_bundle while probing a servicing location, because hammer servicing takes highest priority.
```
Maybe better explanation would be:
"Because servicing locations are never in the bundle" - it's kind of both - servicing location takes precedence over bundle AND it can never be "in the bundle".
I'm probably just getting paranoid :smile:

swaroop-sridhar replied:
I wrote this check thinking about "If someone sets `$CORE_SERVICING` to app directory". But that's very weird and should be explicitly not supported. I'll instead add an assert to check mutual-exclusion of `is_servicing` and `look_in_bundle` cases.

vitek-karas commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/HammerServiceTest.cs:
```diff
@@ -0,0 +1,100 @@
+                    .PublishProject(runtime: TestFixture.CurrentRid, 
+                                    outputDirectory: BundleHelper.GetPublishPath(TestFixture));
+
+                ServiceFixture = new TestProjectFixture("HammerServiceHelper", RepoDirectories, assemblyName: "Location");
```
"HammerServicingHelper" -> "ServicedLocation", or "Location_Serviced" - to make it clear what it's for.

vitek-karas reviewed (approved):
Looks good, thanks!

vitek-karas commented:
@elinor-fung please take a quick look if you get a chance since you're also familiar with this code.

sbomer reviewed (approved):
LGTM though I'm not familiar with this code.

swaroop-sridhar committed:
Update comments in src/installer/corehost/cli/bundle/file_entry.h

Co-authored-by: Vitek Karas <vitek.karas@microsoft.com>

swaroop-sridhar committed:
Update comments in src/installer/corehost/cli/deps_entry.cpp

Co-authored-by: Vitek Karas <vitek.karas@microsoft.com>

swaroop-sridhar committed:
Address Code Review Feedback

swaroop-sridhar commented on src/installer/corehost/cli/deps_entry.cpp:
```diff
@@ -99,6 +100,22 @@ bool deps_entry_t::to_path(const pal::string_t& base, const pal::string_t& ietf_
+        // If a file is resolved to the servicing directory, disable it in the bundle directory.
+        // This step is necessary because runtime resolves assemblies in the bundle even before
+        // attempting to resolve using the TPA. Disabling the file's entry in the bundle
+        // ensures that the servicing entry in the TPA gets priority.
```
@elinor-fung  I agree that the processing in the various `to_path` routines is a bit confusing. 
I've simplified the cases through which `is_servicing` flows a bit -- only through `to_full_path -> to_rel_path -> to_path`. 

A property of the current code is that the probing (whether bundle or file) is done only in `to_path`. 
I'd like to keep it this way, rather than have the caller again repeat the `sub_path` computations and probe/disable the file in bundle. 

I agree that this code can be refactored better, but I'd prefer to not do it in ask-mode. Thanks.

elinor-fung replied:
Okay - can we at least update the description in the comments for the method (line 24 - I can't comment on the line in GitHub) and the header to call out that this function handles this disabling, so can affect broader state in the app now? (A `to_path` const function just sounds so benign relative to what it is actually responsible for now).

swaroop-sridhar commented:
Thanks for the feedback @vitek-karas, @sbomer, @elinor-fung.
I've now addressed the feedback, and made changes except where noted in comments. Thanks.
